### PR TITLE
Common main index handling.

### PR DESCRIPTION
**Pull request by pirogronian**
Currently, Celestia has separate indexing mechanism for stars, dsos and other bodies. This strongly limits possible use of catalogues consisting objects from various categories (and duplicates infrastructure). Moreover, current impelemntation of object index, single sorted vector, have performance limitation in case of large number of objects.

So, this PR provides unified main index infrastructure, enabling cross-indexing for all types of objects. Furthermore, single sorted vector is replaced by array of vectors, which should greately reduce lookup time (and load time on case of really large number of objects, for example > 10000000). 

I think I found reasonable implementation. At least tests with big `stars.dat` from CO showed no considerable increase in startup time and memory usage. Should be fine until `IndexNumber` remains 32 bit... Later some sort of smart hash table will be needed.

### CODE REVIEW

File: src/celengine/astroobj.cpp [hunk 1]
```diff
@@ -5,18 +5,61 @@
 #include "astroobj.h"
 #include "category.h"
 
+AstroObject::AstroObject(AstroObject &o)
+{
+    setIndex(o.getIndex());
```
Comment (hv-work) [inline src/celengine/astroobj.cpp]: you don't need wrapper functions here
just `m_index = o.m_index`

File: src/celengine/astroobj.cpp [hunk 2]
```diff
@@ -5,18 +5,61 @@
...
+        for (auto *cat : *cats)
+            addToCategory(cat);
+}
+
+AstroObject::AstroObject(AstroObject &&o)
+{
+    setIndex(o.getIndex());
+
+    auto cats = o.getCategories();
+    if (cats != nullptr)
```
Comment (hv-work) [inline src/celengine/astroobj.cpp]: and this block is redundant, all you need is `m_categories = o.m_categories; o.m_categories = nullptr;`

File: src/celengine/astroobj.cpp [hunk 3]
```diff
@@ -5,18 +5,61 @@
...
+        for (auto *cat : *cats)
+            addToCategory(cat);
+}
+
+AstroObject::AstroObject(AstroObject &&o)
+{
+    setIndex(o.getIndex());
```
Comment (hv-work) [inline src/celengine/astroobj.cpp]: o.m_index = 0;

File: src/celengine/dsooctree.cpp [hunk 4]
```diff
@@ -43,7 +43,7 @@ double dsoAbsoluteMagnitudeDecayFunction(const double excludingFactor)
 
 
 template <>
-DynamicDSOOctree* DynamicDSOOctree::getChild(DeepSkyObject* const & _obj, const PointType& cellCenterPos)
+DynamicDSOOctree* DynamicDSOOctree::getChild(DeepSkyObject* & _obj, const PointType& cellCenterPos)
```
Comment (hv-work) [inline src/celengine/dsooctree.cpp]: reference to pointer?

Comment (pirogronian) [reply src/celengine/dsooctree.cpp]: Well, the only thing I did was to remove `const`. For all the rest, ask original author, please :-)

File: src/celengine/astroobj.cpp (hunk 2 above)
Comment (pirogronian) [reply src/celengine/astroobj.cpp]: Not exactly. Categories would't know the object is moved. This would lead to inconsistence.

File: src/celengine/astroobj.cpp (hunk 1 above)
Comment (pirogronian) [reply src/celengine/astroobj.cpp]: Oh yes, we need. See at implementation of `setIndex`.

File: src/celengine/astroobj.cpp (hunk 3 above)
Comment (pirogronian) [reply src/celengine/astroobj.cpp]: Why "0"?? And see my above comment.

**Comment by pirogronian** [comment]
@hv-work
Btw, I requested review of #651 . This one is unfinished, but thanks anyway :-)

File: src/celengine/astroobj.cpp (hunk 2 above)
Comment (375gnu) [reply src/celengine/astroobj.cpp:16]: in that case why not to remove reference to old object and add one to new object?

**Comment by 375gnu** [comment]
i actually don't understand what is going here.

could you provide a problem description and how your pr fixes this problem?

**Comment by pirogronian** [comment]
General idea or technical details?
I put general idea to the PR description. Several index-related methods in AstroObject is a result of seeking performance gain during data loading, mainly to avoid unnecessary lookup for a duplicate index, if we are sure there is no such a one. Maybe describe every method in doxygen-like style?

**Comment by 375gnu** [comment]
> Maybe describe every method in doxygen-like style?

This would be great.

File: src/celengine/octree.h [hunk 5]
```diff
@@ -300,9 +300,9 @@ inline void DynamicOctree<OBJ, PREC>::rebuildAndSort(StaticOctree<OBJ, PREC>*& _
     OBJ* _firstObject = _sortedObjects;
 
     if (_objects != nullptr)
-        for (typename ObjectList::const_iterator iter = _objects->begin(); iter != _objects->end(); ++iter)
+        for (typename ObjectList::iterator iter = _objects->begin(); iter != _objects->end(); ++iter)
```
Comment (375gnu) [inline src/celengine/octree.h:303]: could you rewrite this using C++11 style?

File: src/celengine/octree.h [hunk 6]
```diff
@@ -300,9 +300,9 @@ inline void DynamicOctree<OBJ, PREC>::rebuildAndSort(StaticOctree<OBJ, PREC>*& _
...
         {
-            *_sortedObjects++ = **iter;
+            new (_sortedObjects++) OBJ(std::move(**iter));
```
Comment (375gnu) [inline src/celengine/octree.h:305]: and move `_sortedObjects++` out of `new`, because of UB you might have here either `_sortedObjects` or `_sortedObjects + 1` if i correctly understand all this postfix `++` crap.

File: src/celengine/star.cpp [hunk 7]
```diff
@@ -945,6 +945,13 @@ StarDetails::setInfoURL(const string& _infoURL)
 }
 
 
+Star::Star(Star &&s) : AstroObject(std::move(s))
+{
+    setPosition(s.getPosition());
+    setAbsoluteMagnitude(s.getAbsoluteMagnitude());
+    setDetails(s.getDetails());
```
Comment (375gnu) [inline src/celengine/star.cpp:952]: please use direct access to private members

File: src/celengine/star.h [hunk 8]
```diff
@@ -241,8 +241,12 @@ class Star : public AstroObject
     Star() = default;
     virtual ~Star();
 
+    Star(Star &&s);
+
```
Comment (375gnu) [inline src/celengine/star.h:245]: add `Star(const Star&)` and appropriate operators `operator=`

File: src/celengine/stardb.cpp [hunk 9]
```diff
@@ -691,28 +680,17 @@ bool StarDatabase::loadBinary(istream& in)
 
         nStars++;
     }
+    size_t i = 0;
+    while(i < nStars)
```
Comment (375gnu) [inline src/celengine/stardb.cpp:684]: please use `for` loop

File: src/celengine/stardb.cpp [hunk 10]
```diff
@@ -754,7 +732,7 @@ void StarDatabase::finish()
 static void stcError(const Tokenizer& tok,
                      const string& msg)
 {
-    fmt::fprintf(cerr,  _("Error in .stc file (line %i): %s\n"), tok.getLineNumber(), msg);
+    fmt::fprintf(cout,  _("Error in .stc file (line %i): %s\n"), tok.getLineNumber(), msg);
```
Comment (375gnu) [inline src/celengine/stardb.cpp:735]: in final version please revert this change

File: src/celengine/stardb.cpp [hunk 11]
```diff
@@ -980,7 +958,7 @@ bool StarDatabase::createStar(Star* star,
 
                 if (!hasBarycenter)
                 {
-                    fmt::fprintf(cerr, _("Barycenter %s does not exist.\n"), barycenterName);
+                    fmt::fprintf(cout, _("Barycenter %s does not exist.\n"), barycenterName);
```
Comment (375gnu) [inline src/celengine/stardb.cpp:961]: as above

File: src/celengine/stardb.cpp [hunk 12]
```diff
@@ -1313,12 +1283,13 @@ bool StarDatabase::load(istream& in, const fs::path& resourcePath)
         bool ok = false;
         if (isNewStar && disposition == DataDisposition::Modify)
         {
-            clog << "Modify requested for nonexistent star.\n";
+            cout << "Modify requested for new star.\n";
```
Comment (375gnu) [inline src/celengine/stardb.cpp:1286]: as above, revert cout to clog

File: src/celutil/arraymap.h [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+        return k >> (sizeof(k) * 8 - ArrayKeyLen);
+    }
+    size_t size() const { return ArraySize; }
+protected:
+    V m_array[ArraySize];
+    size_t m_used { 0 };
+public:
+    ArrayMap()
+    {
+        for(size_t i = 0; i < ArraySize; i++)
```
Comment (375gnu) [inline src/celutil/arraymap.h:26]: use `std::fill` instead

File: src/celutil/packedrangeset.h [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+            if (isSorted() && k < getKey(m_vector[i]))
+                return -1;
+        }
+        return -1;
+    }
+    bool insert(const V &v)
+    {
+        if (getSize() > 0 && isWithinRange(getKey(v)))
+        {
+            for (typename Container::iterator it = m_vector.begin(); it != m_vector.end(); it++)
```
Comment (375gnu) [inline src/celutil/packedrangeset.h:139]: `auto` ?

File: src/celengine/star.h (hunk 8 above)
Comment (pirogronian) [reply src/celengine/star.h:245]: Should I? AstroObject doesn't have copy constructor either. I think objects shouldn't be copy-able.

File: src/celengine/stardb.cpp (hunk 10 above)
Comment (pirogronian) [reply src/celengine/stardb.cpp:735]: I was sure I already reverted it... :open_mouth:

File: src/celengine/octree.h (hunk 5 above)
Comment (pirogronian) [reply src/celengine/octree.h:303]: I got
> src/celengine/dsodb.cpp:385:46:   required from here
> src/celengine/octree.h:303:9: error: no matching function for call to ‘begin(std::vector<DeepSkyObject**, std::allocator<DeepSkyObject**> >*&)’

File: src/celengine/star.cpp (hunk 7 above)
Comment (pirogronian) [reply src/celengine/star.cpp:952]: Why You prefer direct access? Using accessor seems to be more stable.

File: src/celengine/star.h (hunk 8 above)
Comment (375gnu) [reply src/celengine/star.h:245]: in this case delete the explicitly:
```
Star(const Star&) = delete;
```
I'm sorry but C++ requires this.

File: src/celengine/star.cpp (hunk 7 above)
Comment (375gnu) [reply src/celengine/star.cpp:952]: outside of the class - yes. but not here. first, this is internal class implementation. second it must set moved object's properties to "null" (nullptr for pointers, etc), you should not duplicate or add validation layers here, just take there, put here and make destructor called on the moved object essentially "nop".

File: src/celengine/octree.h (hunk 5 above)
Comment (375gnu) [reply src/celengine/octree.h:303]: for `for (const auto &o : *_objects)`?

File: src/celengine/dsooctree.cpp (hunk 4 above)
Comment (375gnu) [reply src/celengine/dsooctree.cpp:46]: if you change something, cleanup please

File: src/celengine/star.cpp (hunk 7 above)
Comment (pirogronian) [reply src/celengine/star.cpp:952]: Ok, essentially understood.

File: src/celengine/octree.h (hunk 5 above)
Comment (pirogronian) [reply src/celengine/octree.h:303]: Ah, forgot that `_objects` is a pointer...

File: src/celengine/star.h (hunk 8 above)
Comment (pirogronian) [reply src/celengine/star.h:245]: Ok. But why not to delete it in `AstroObject` only? Should I delete it in every derived class?

File: src/celengine/astroobj.cpp (hunk 2 above)
Comment (pirogronian) [reply src/celengine/astroobj.cpp:16]: Hm... It could be a bit more efficient, but modification of elements inside set will change its order, hence will make it decoherent. It could have a point for map or vector, but is it worth effort?

File: src/celengine/astroobj.cpp (hunk 3 above)
Comment (pirogronian) [reply src/celengine/astroobj.cpp]: Ok, you mean InvalidIndex.

File: src/celengine/dsooctree.cpp (hunk 4 above)
Comment (pirogronian) [reply src/celengine/dsooctree.cpp:46]: General implementation is done for references, while specialisation for DSOs uses pointers. It wouldn't be cleanup, but rewrite.

File: src/celengine/star.h (hunk 8 above)
Comment (375gnu) [reply src/celengine/star.h:245]: derived classes should derive all constructors, and deleted too

**Comment by 375gnu** [comment]
Memory usage is 188MB vs 158MB. The growth is too large. But I haven't tested with CO yet.

**Comment by pirogronian** [comment]
I'm surprised, as minimal memory usage growth should be 32MB :-). But it's constant growth, independent of objects number. If it's inacceptable anyway, I can tune container further.

**Comment by pirogronian** [comment]
@375gnu
I'm waiting for your final review. Should I wait a bit more or you are just waiting until I improve minimal memory usage?

**Comment by 375gnu** [comment]
With CO stars.dat I have 454MB vs 363MB. If I increase limiting magnitude to 12 then it lifts to 470MB and even after returning magnitude to 8 it doesn't decrease. I'm afraid you need to optimize it.

**Comment by pirogronian** [comment]
It's unbelivable, as my code allocates memory only for loading data at startup and has nothing to do with rendering. Moreover, I'm unable to reproduce this runtime memory usage increase.

**Comment by pirogronian** [comment]
Adjusted a bit. But in practise memory efficiency still strongly depend on numbers distribution.

**Comment by 375gnu** [comment]
167MB with the default stars.dat with growth to 177 (max magnitude + position on Milky way).
447MB with CO's stras.dat with growth to 459

File: src/celutil/hashmap.h [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+    bool m_hasZero { false };
+public:
+    static V invalidValue();
+    size_t size() const { return m_size + 1; }
+    size_t used() const { return m_hasZero ? m_used + 1 : m_used; }
+    const K *keyData() const { return m_keys; }
+    const V *valData() const { return m_vals; }
+    size_t hashedKey(K k) const
+    {
+        return (k ^ k >> (sizeof(k) * 4) * 0x9E3779B97F4A7C15L >> m_shift);
```
Comment (375gnu) [inline src/celutil/hashmap.h:30]: where this magic came from?

Comment (pirogronian) [reply src/celutil/hashmap.h:30]: Adapted from https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/utils/LongMap.java
When in single function, gave wrong results.
Anyway, I'm again out of developing. Not sure, if I'll return again. You can forget this PR, if you have better things to do.

Comment (375gnu) [reply src/celutil/hashmap.h:30]: that's sad

**Comment by 375gnu** [comment]
closing as no activities
